cmd/cluster-init/update_bootstrap_secrets.go

```diff
@@ -0,0 +1,352 @@
+package main
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	"github.com/openshift/ci-tools/pkg/api/secretbootstrap"
+)
+
+const (
+	credentials         = "credentials"
+	regPullCredsAll     = "registry-pull-credentials-all"
+	dotDockerConfigJson = ".dockerconfigjson"
+	testCredentials     = "test-credentials"
+	kubeconfig          = "kubeconfig"
+	config              = "config"
+)
+
+type pushPull string
+
+const (
+	pull pushPull = "puller"
+	push pushPull = "pusher"
+)
+
+func updateCiSecretBootstrap(o options) error {
+	secretBootstrapDir := filepath.Join(o.releaseRepo, "core-services", "ci-secret-bootstrap")
+	secretBootstrapConfigFile := filepath.Join(secretBootstrapDir, "_config.yaml")
+	logrus.Infof("Updating ci-secret-bootstrap: %s", secretBootstrapConfigFile)
+
+	var c secretbootstrap.Config
+	if err := secretbootstrap.LoadConfigFromFile(secretBootstrapConfigFile, &c); err != nil {
+		return err
+	}
+	if err := updateCiSecretBootstrapConfig(o, &c); err != nil {
+		return err
+	}
+	return secretbootstrap.SaveConfigToFile(secretBootstrapConfigFile, &c)
+}
+
+func updateCiSecretBootstrapConfig(o options, c *secretbootstrap.Config) error {
+	for _, groupName := range []string{buildUFarm, "non_app_ci", "non_app_ci_x86"} {
+		c.ClusterGroups[groupName] = append(c.ClusterGroups[groupName], o.clusterName)
+	}
+	c.UserSecretsTargetClusters = append(c.UserSecretsTargetClusters, o.clusterName)
+
+	for _, step := range []func(c *secretbootstrap.Config, o options) error{
+		updatePodScalerSecret,
+		updateBuildFarmSecrets,
+		updateDPTPControllerManagerSecret,
+		updateRehearseSecret,
+		updateChatBotSecret,
+		updateExistingRegistryPullCredentialsAllSecrets,
+	} {
+		if err := step(c, o); err != nil {
+			return err
+		}
+	}
+
+	appendSecret(generateRegistryPushCredentialsSecret, c, o)
+	appendSecret(generateRegistryPullCredentialsSecret, c, o)
+	appendSecret(generateCiOperatorSecret, c, o)
+	generateRegistryPullCredentialsAllSecrets(c, o)
+	return nil
+}
+
+func appendSecret(secretGenerator func(options) secretbootstrap.SecretConfig, c *secretbootstrap.Config, o options) {
```

Comment:
This would make all of your steps homogenous, so you have just the one loop:


```go
func appendSecret(secretGenerator func(options) secretbootstrap.SecretConfig) func(c *secretbootstrap.Config, o options) error {
	return func(c *secretbootstrap.Config, o options) error{
		secret := secretGenerator(o)
		logrus.Infof("Creating new secret with 'to' of: %v", secret.To)
		c.Secrets = append(c.Secrets, secret)
		return nil
	}
}
```

Reply:
ah, that is nicer